File: jqtype-ts/src/diagnostic.ts
export type Severity = "Warning" | "Error";

export interface SourceSpan {
  start: number;
  end: number;
}

export interface Diagnostic {
  severity: Severity;
  message: string;
  span: SourceSpan | null;
  source_name: string | null;
}

export const SourceSpan = {
  new(start: number, end: number): SourceSpan {
    return { start, end };
  },
};

export const Diagnostic = {
  warning(message: string, span: SourceSpan | null): Diagnostic {
    return { severity: "Warning", message, span, source_name: null };
  },

  error(message: string, span: SourceSpan | null): Diagnostic {
    return { severity: "Error", message, span, source_name: null };
  },

  withSourceName(diagnostic: Diagnostic, source_name: string | null): Diagnostic {
    return { ...diagnostic, source_name };
  },
};

File: jqtype-rs/crates/jqtype-core/src/diagnostic.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum Severity {
    Warning,
    Error,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

impl SourceSpan {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    pub fn from_range(range: core::ops::Range<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Diagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Option<SourceSpan>,
    pub source_name: Option<String>,
}

impl Diagnostic {
    pub fn warning(message: impl Into<String>, span: Option<SourceSpan>) -> Self {
        Self {
            severity: Severity::Warning,
            message: message.into(),
            span,
            source_name: None,
        }
    }

    pub fn error(message: impl Into<String>, span: Option<SourceSpan>) -> Self {
        Self {
            severity: Severity::Error,
            message: message.into(),
            span,
            source_name: None,
        }
    }

    pub fn with_source_name(mut self, source_name: Option<String>) -> Self {
        self.source_name = source_name;
        self
    }
}
